Prepare for 1.11.2 release

LordAro opened the pull request:
## Motivation / Problem
Keep getting reports for OGL related crashes

## Description
Backported almost everything that's been merged as of 13:25UTC 02/05
A few needed manual work to backport. Network stuff is getting tricky to untangle :)

## Limitations
~Still waiting on a few PRs~

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

glx22 committed:
Fix: Auto-fill version details in rev.cpp and ottres.rc

glx22 committed:
Fix 799eb31: [CMake] Don't include regression AIs in bundles (#9068)

TrueBrain committed:
Doc: update the supported platforms to match current master

TrueBrain committed:
Doc: fix OpenGFX / OpenSFX / OpenMSX links in README

TrueBrain committed:
Doc: split up which libraries are used for which situation

TrueBrain committed:
Doc: indicate the most common options for CMake

michicc committed:
Fix: [OpenGL] Check maximum supported texture size against screen resolution.

michicc committed:
Codechange: [Win32] Try getting an OpenGL 4.5 context first before aiming at 3.2.

michicc committed:
Change: [Win32] Limit the OpenGL video driver to OpenGL 3.2 or newer on Windows.

PeterN committed:
Codechange: Scale rating minigraphs on station window list. (#9075)

Minigraphs did not adjust size to accomodate large text, either by font size or font zoom, leading to cropped labels.
Minigraphs and spacing are now scaled by font zoom, as this seems to behave better than gui zoom in this instance.

glx22 committed:
Fix: [CMake] Incorrect dependency checks for GRF

glx22 committed:
Fix: [CMake] Check nforenum and grfcodec return value

PeterN committed:
Fix #9097: Upper 16 bits of cargo base payment rate were discarded. (#9098)

NewGRF spec says that base payment rate is 32 bits, but it was loaded into a 16 bit variable. This value is loaded into Money variable after inflation is applied.

michicc committed:
Fix: [OpenGL] Main loop expects to start with the video buffer unmapped. (#9100)

rubidium42 committed:
Fix: [Network] Prevent stalling save game transfer when compression is slow

PeterN committed:
Fix: Use width of tiny arrow string instead of scaled pixels in ship/aircraft list. (#9102)

PeterN committed:
Fix: Count engine details text in lines rather than pixels. (#9107)

This allows the details panel to scale correctly for different zoom levels.

TrueBrain committed:
Codechange: move all NetworkGameInfo related functions to a single file

It currently was a bit scattered over the place. Part of
NetworkGameInfo is also the GRF Identifiers that goes with it.

TrueBrain committed:
Add: ability to retrieve game info from server over TCP

rubidium42 committed:
Fix: [Network] errno and strerror do not handle network errors on Windows

rubidium42 committed:
Codechange: [Network] Do not leak os_abstraction.h via network_func

rubidium42 committed:
Codechange: [Network] Do not leak os_abstraction.h via fios.h

rubidium42 committed:
Codechange: [Network] Do not leak os_abstraction.h via base_media_func.h

rubidium42 committed:
Change: [Network] Safeguard from using errno/strerror for handling network errors

They are likely not working as expected on Windows, so prevent their usage.
Winsock does not set errno and strerror does not return anything useful for Winsock error numbers.

rubidium42 committed:
Change: [Network] Use string error messages instead of numeric error numbers that need to be looked up

TrueBrain committed:
Codechange: be explicit in pointer comparisons

Milek7 committed:
Codechange: Use __attribute__ access none to silence GCC 11 -Wmaybe-uninitialized warnings (#9124)

TrueBrain committed:
Fix b3003dd1: swap SERVER_GAME_INFO with CLIENT_GAME_INFO (#9129)

The idea is that if you query an older server that does not support
this packet yet, the client receives an error. The assumption was
that on every "illegal packet" the connection would be closed. This
turns out to be false.

Now CLIENT_GAME_INFO aligns with the old PACKET_CLIENT_NEWGRFS_CHECKED,
which does a pre-check (which fails), and an error is sent back
and the connection is closed.

This is not a nice solution, but it is the best we got.

PeterN committed:
Fix: Incorrect vertical alignment of icon and text in DropDownListIconItem. (#9133)

This happens if the bounding dimensions are changed so that each item is the same size, as happens on the railtype/roadtype dropdown lists, as the vertical offset was calculated before this dimension is changed.

rubidium42 committed:
Fix: String validation could leave invalid Utf8 encoded strings (#9096)

In case a character was encoded in multiple bytes, but required fewer bytes to be encoded, the first byte would be copied to the output leaving an invalid Utf8 encoded string. Later uses of the validated string would use the same decode logic, which would yield a question mark and just read a single byte, so nothing dangerous happened.
Furthermore, because the next byte would not be a first byte of an encoded Utf8 character, the last few valid characters could be removed by the validation as well.

TrueBrain committed:
Remove: performance measurements in YAPF

YAPF was constantly measuring its performance, but only at
certain debug-levels this information was shown.

Now after years, I sincerely wonder if anyone still knows about this
feature and who still use it. Especially with the new framerate window,
this detailed performance is not as meaningful anymore as it once
was.

TrueBrain committed:
Cleanup: remove weird left-over comment in yapf.hpp

rubidium42 committed:
Codechange: encapsulate network error handling

rubidium42 committed:
Codechange: rename NetworkError to ShowNetworkError

rubidium42 committed:
Codechange: move some OS abstraction method implementations out of the header

rubidium42 committed:
Fix #9152, Fix #9153: screenshot command showed error messages when successful

glx22 committed:
Fix d4f0b6f4: [CMake] CMAKE_PROJECT_VERSION_XXX are not in CMake 3.9 (#9154)

LordAro committed:
Update: Changelog for 1.11.2 release

rubidium42 committed:
Fix #6598: Do not disconnect before company number validation

NetworkClientConnectGame already does a NetworkDisconnect, so no reason to do it here

rubidium42 committed:
Fix #6598: Prevent invalid memory accesses when abandoning a join from within a network game

One could join a network game from within an already running network game. This would call a NetworkDisconnect, but keeps the UI alive. If, during that process the join is aborted, e.g. by cancelling on a password dialog, you would still be in your network game but also get shown the server list.
Solve all the underlying problems by falling back to the main UI when (re)connecting to a(nother) server.

glx22 committed:
Change: [Actions] Add a 2 minutes timeout for regression test (#9166)

glx22 committed:
Fix: Don't consider regression AIs when starting a random AI (#9164)

michicc committed:
Codechange: Generalise the delayed blitter change to a generic video driver command queue.

michicc committed:
Fix #9147: Delay making screenshots until the next draw tick as we may not access the video buffer from the game thread.

frosch123 committed:
Fix: [NewGRF] industry variables 65 and 66 ignored the parameter, and always used the north tile.

frosch123 committed:
Fix: [NewGRF] industry variable 66 and object variable 46 clamped the squared-euclidian distance to 16 bit, when they should not.

rubidium42 committed:
Fix: [Network] Reading beyond the length of the server's ID when hashing password

Under normal circumstances the server's ID is 32 characters excluding '\0', however this can be changed at the server. This ID is sent to the server for company name hashing. The client reads it into a statically allocated buffer of 33 bytes, but fills only the bytes it received from the server. However, the hash assumes all 33 bytes are set, thus potentially reading uninitialized data, or a part of the server ID of a previous game in the hashing routine.
It is still reading from memory assigned to the server ID, so nothing bad happens, except that company passwords might not work correctly.

PeterN committed:
Fix: Crash when extra viewport height is zero with sign in view. (#9175)

If a viewport sign straddles the top of a viewport, a crash will occur if the viewport height is zero. This is resolved by simply not attempting to draw the viewport in this situation, consistent with other widgets.

PeterN committed:
Fix: Update text effect size when font zoom is changed. (#9174)

glx22 commented on CMakeLists.txt:
```diff
@@ -4,7 +4,9 @@ if(NOT BINARY_NAME)
 
-project(${BINARY_NAME})
+project(${BINARY_NAME}
+    VERSION 1.11.1
```
Should become 1.11.2 at some point.

rubidium42 commented on changelog.txt:
```diff
@@ -1,3 +1,21 @@
+------------------------------------------------------------------------
+Change: [Win32] Limit hardware accelerated video driver to OpenGL 3.2 or higher (#9077)
+Change: More improvements to the GUI at different scales (#9075, #9102, #9107, #9133, #9174)
+Fix: [Network] Possible out-of-bounds memory access with certain password lengths (#9176)
```
It's actually part of the salt used for company passwords where the out of bounds happens on. It has nothing to do with the actual length of the passwords.
What about: `Fix: [Network] Possible out-of-bounds memory access with certain lengths of server supplied company password salts (#9176)`

rubidium42 commented on changelog.txt:
```diff
@@ -1,3 +1,21 @@
+Fix: [NewGRF] Industry variables 65 and 66 ignored the parameter, and always used the north tile (#9088)
+Fix: Do not include regression test AI in bundle (#9068)
+Fix #9062: Executable version was not updated (#9066, #9154)
+
```
Have the following been left consciously out of the changelog?
* Fix: String validation could leave invalid Utf8 encoded strings (#9096)
* Fix #6598: [Network] Do not disconnect from network game before company number validation (#9163)
* Fix #6598: [Network] Fall back to main menu when joining a network game to prevent invalid game state (#9163)

Milek7 committed:
Codechange: Acquire video buffer before taking game state lock to prevent erratic fast forward behaviour (#9140)

LordAro committed:
Update: Backport language changes

PeterN commented:
Please include #9183 and #9182

JGRennison committed:
Fix #9113: Assertion failure when removing airport with order backup (#9182)

PeterN committed:
Fix #9174: Don't update text effect if it has been reset. (#9183)

PeterN committed:
Fix: Query windows may be partially drawn initially. (#9184)

Query window was not marked dirty after being moved on init. It was then marked dirty once the white border flash completed.

rubidium42 committed:
Fix #9117, 04ce1f07: [Fluidsynth] Infinite wait when stopping song (#9181)

In FluidSynth 2.2.0 an extra state was added to denote stopping. To transition
from this state to a stopped state the rendering needs to be running. Since
04ce1f07 locking was added that skipped the rendering when something else held
a lock, so the state would never get to stopped and join would never return.

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+------------------------------------------------------------------------
+Change: [Win32] Limit hardware accelerated video driver to OpenGL 3.2 or higher (#9077)
+Change: More improvements to the GUI at different scales (#9075, #9102, #9107, #9133, #9174, #9183)
+Fix: Query windows could be partially drawn initially (#9184)
```
this sentence is hard to read. Mainly, I think, because "initially" is at the end. Maybe just remove that word completely?

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Change: [Win32] Limit hardware accelerated video driver to OpenGL 3.2 or higher (#9077)
+Change: More improvements to the GUI at different scales (#9075, #9102, #9107, #9133, #9174, #9183)
+Fix: Query windows could be partially drawn initially (#9184)
+Fix #9113: Potential crash when removing an airport that exists in an aircraft's orders (#9182)
```
You use "potential" a few times .. I think it a weird way to describe a bug and/or fix. Like "we are not sure this could happen, but if it did, it no longer does". Pretty sure that is often not the case ;)

I think we should just remove "potential" everywhere, before every line is infested with that word :D

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix: Query windows could be partially drawn initially (#9184)
+Fix #9113: Potential crash when removing an airport that exists in an aircraft's orders (#9182)
+Fix #9117: [Fluidsynth] Potential hang when changing song (#9181)
+Fix: String validation could leave invalid UTF8 encoded strings (#9096)
```
nit: it is `UTF-8`

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix #9113: Potential crash when removing an airport that exists in an aircraft's orders (#9182)
+Fix #9117: [Fluidsynth] Potential hang when changing song (#9181)
+Fix: String validation could leave invalid UTF8 encoded strings (#9096)
+Fix: [Network] Possible out-of-bounds memory access with certain lengths of server supplied company password salts (#9176)
```
Suggestion:
`Fix: [Network] Out-of-bounds memory access with modified servers sending too short salts`

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix #9147: Potential crash when taking screenshots (#9169)
+Fix #6598: [Network] Prevent potential crashes when (re)joining network game by falling back to main menu first (#9163)
+Fix #9152: Screenshot success popup window was treated as an error (#9159)
+Fix: Fast-forward behaviour when vsync is enabled (#9140)
```
This is also weird .. what is the fix?
`Fix: fast-forward stuttering when vsync enabled`
maybe?

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix #6598: [Network] Prevent potential crashes when (re)joining network game by falling back to main menu first (#9163)
+Fix #9152: Screenshot success popup window was treated as an error (#9159)
+Fix: Fast-forward behaviour when vsync is enabled (#9140)
+Fix: [Network,Win32] Network errors were handled badly (#9116)
```
missing space between `Network` and `Win32`.

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix #9152: Screenshot success popup window was treated as an error (#9159)
+Fix: Fast-forward behaviour when vsync is enabled (#9140)
+Fix: [Network,Win32] Network errors were handled badly (#9116)
+Fix: [Network] Savegame transfer could stall in certain circumstances (#9106)
```
Another way of saying "potential" :D

Suggestion: `Fix: [Network] Savegame transfer could stall in rare cases`

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix: Fast-forward behaviour when vsync is enabled (#9140)
+Fix: [Network,Win32] Network errors were handled badly (#9116)
+Fix: [Network] Savegame transfer could stall in certain circumstances (#9106)
+Fix #9097: [NewGRF] Cargo variable initial payment variable was being truncated (#9098)
```
one too many `variable` in the sentence I think?

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix #9097: [NewGRF] Cargo variable initial payment variable was being truncated (#9098)
+Fix: [NewGRF] Industry variable 66 and object variable 46 erroneously truncated the distance (#9088)
+Fix: [NewGRF] Industry variables 65 and 66 ignored the parameter, and always used the north tile (#9088)
+Fix: Do not include regression test AI in bundle (#9068)
```
Maybe combine this with #9164. If we don't ship it, that you cannot use it is no longer relevant?

TrueBrain commented on changelog.txt:
```diff
@@ -1,3 +1,27 @@
+Fix: [NewGRF] Industry variable 66 and object variable 46 erroneously truncated the distance (#9088)
+Fix: [NewGRF] Industry variables 65 and 66 ignored the parameter, and always used the north tile (#9088)
+Fix: Do not include regression test AI in bundle (#9068)
+Fix #9062: Executable version was not updated (#9066, #9154)
```
Suggestion: `Fix #9062: [Windows] version in executable was not set to current release version`

TrueBrain commented:
Did some smoke-testing, all seems to work fine.

I also tested if our addition to the network protocol works as expected, and it does: `master` clients can query 1.11.2 servers via TCP while 1.11.2 clients still do it over UDP; both work as expected.

So as far as I can tell, all your backporting efforts are a success :) \o/
